File: client/src/app/shared/Models/Order.ts
import {IAddress} from "./address";

export interface IOrder {
  id: number;
  buyerEmail: string;
  orderDate: Date;
  shipToAddress: IAddress;
  deliveryMethod: string;
  shippingPrice: number;
  orderItems: IOrderItem[];
  subtotal: number;
  status: string;
  total: number;
}

export interface IOrderItem {
  productId: number;
  productName: string;
  picturUrl: string;
  price: number;
  quantity: number;
}



export enum OrderStatus {
Pending = "Pending",
PaymentRecevied = "PaymentRecevied",
PaymentFailed = "PaymentFailed",
InProgress  = "InProgress",
InShipping  = "InShipping",
Delivered = "Delivered",
Canceled = "Canceled",
}

export let OrderStatusDisplay:Record<OrderStatus, string> = {
  [OrderStatus.Pending]: "Pending",
  [OrderStatus.PaymentRecevied]: "Payment Recevied",
  [OrderStatus.PaymentFailed]: "Payment Failed",
  [OrderStatus.InProgress]: "In Progress",
  [OrderStatus.InShipping]: "In Shipping",
  [OrderStatus.Delivered]: "Delivered",
  [OrderStatus.Canceled]: "Canceled",
}

File: ECommerce/Models/Order/Order.cs
namespace ECommerce.Models.Order
{
    public class Order : BaseEntity
    {
        public Order()
        {

        }

        public Order(Address shipToAddress, DelivaryMethod delivaryMethod, IReadOnlyList<OrderItem> orderItems, string buyerEmail, decimal subtotal)
        {
            BuyerEmail = buyerEmail;
            ShipToAddress = shipToAddress;
            DelivaryMethod = delivaryMethod;
            OrderItems = orderItems;
            Subtotal = subtotal;
        }

        public string BuyerEmail { get; set; }
        public DateTimeOffset OrderDate { get; set; } = DateTimeOffset.Now;
        public Address ShipToAddress { get; set; }
        public DelivaryMethod DelivaryMethod { get; set; }
        public IReadOnlyList<OrderItem> OrderItems { get; set; }
        public decimal Subtotal { get; set; }
        public OrderStatus Status { get; set; } = OrderStatus.Pending;
        public string? PaymentIntentId { get; set; }
        public string? CanceledReason { get; set; }
        
        public decimal GetTotal()
        {
            return Subtotal + DelivaryMethod.Price;
        }
    }
}
